packages/cli/src/schema/custom-validations/moving-averages.ts

```diff
@@ -0,0 +1,94 @@
+import set from 'lodash/set';
+import { isDefined } from 'ts-is-present';
+
+function hasValuesProperty(
+  input: [string, { values?: Record<string, unknown>[] }]
+): input is [string, { values: Record<string, unknown>[] }] {
+  return input[1].hasOwnProperty('values');
+}
+
+export function validateMovingAverages(
+  input: Record<string, { values?: Record<string, unknown>[] }>
+) {
+  const result = Object.entries(input)
+    // first filter out all the non-metric properties (we only want the ones with a values collection)
+    .filter(hasValuesProperty)
+    // Then filter out all the metrics that contain no properties with a '_moving_average' suffix
+    .filter(([_p, value]) => hasMovingAverages(value.values[0]))
+    // Map to objects that ONLY contain the moving average properties
+    .map(([propertyName, value]) => ({
+      [propertyName]: value.values.map((x: Record<string, unknown>) => {
+        const movingAverageProperties = findMovingAverages(x);
+        return movingAverageProperties.reduce(
+          (aggr, p) => set(aggr, p, x[p]),
+          {} as Record<string, unknown>
+        );
+      }),
+    }))
+    // Perform the validations on those moving averages
+    .map((x) =>
+      Object.entries(x).map(([propertyName, values]) => {
+        const propsWithValuesInFirstSixEntries = hasValuesInFirstSixEntries(
+          values
+        );
+        if (propsWithValuesInFirstSixEntries.length) {
+          return `${propertyName} has non NULL values in the first six items of metric ${propsWithValuesInFirstSixEntries.join(
+            ' and '
+          )}`;
+        }
+        const propsWithNonConsecutiveValues = hasNonConsecutiveValues(values);
+        if (propsWithNonConsecutiveValues.length) {
+          return `${propertyName} has non consecutive values in metric ${propsWithNonConsecutiveValues.join(
+            ' and '
+          )}`;
+        }
+        return undefined;
+      })
+    )
+    .flat()
+    .filter(isDefined);
+
+  return result.length ? result : undefined;
+}
+
+function hasNonConsecutiveValues(values: Record<string, unknown>[]) {
+  const propertyNames = Object.keys(values[0]);
+  return propertyNames.filter((propertyName) =>
+    hasNonConsecutiveValuesInMetric(propertyName, values)
+  );
+}
+
+function hasNonConsecutiveValuesInMetric(
+  propertyName: string,
+  collection: Record<string, unknown>[]
+) {
+  const values = collection.slice(6).map((x) => x[propertyName]);
+  let lastValue = values.pop();
+
+  while (lastValue === null && values.length) {
+    lastValue = values.pop();
+  }
+
+  return values.indexOf(null) > -1;
+}
```

Comment:
It looks like `while (lastValue === null` will not be true if the 7th value is non-null? And then the loop will never be executed.

I think it's unnecessary to mutate and iterate the full array. You can use an early-return statement.

```suggestion
function hasNonConsecutiveValuesInMetric(
  propertyName: string,
  collection: Record<string, unknown>[]
) {
  const values = collection.slice(6).map((x) => x[propertyName]);
  let lastValue;

  for (const value of values) {
     if(lastValue === null && value !== null) {
        return true
     }
  
    lastValue = value
   }
   return false
}
```

Reply:
Your assumption is incorrect, but your code suggestion simplifies things even more, so I've made the update